### ISSUE TITLE
Converting Nordic Bert to Huggingface Transformer BertModel?

### ISSUE THREAD

**Issue by kasperjunge** [mollerhoj/danish_bert#15]
I am trying to convert the danish Version 2 BertModel to a PyTorch model. More specifically I am trying to convert it into the BertModel in Huggingface's transformer library. Does anyone know how to do that?

![billede](https://user-images.githubusercontent.com/39537120/91154410-ca941380-e6c1-11ea-9771-bdd2009a88b2.png)

**Comment by christianbv** [comment]
Where did you find the code on the left hand side? 
I am trying the same, just on the norwegian version. Does the danish vocab file start have ## before all words and ##_ before all "sub-words", kind of like sentencePiece tokenization? Appears as if the norwegian one has the same

**Comment by christianbv** [comment]
Found the code btw. If you just want to convert from tensorflow checkpoints into pytorch then you can just do something like this:



`from transformers import BertConfig, BertForPreTraining, load_tf_weights_in_bert`

`def convert_to_pytorch(tf_checkpoint_path, bert_config_file, pytorch_dump_path):`
    
    # Initialise PyTorch model
    config = BertConfig.from_json_file(bert_config_file)
    print("Building PyTorch model from configuration: {}".format(str(config)))
    model = BertForPreTraining(config)

    # Load weights from tf checkpoint
    load_tf_weights_in_bert(model, config, tf_checkpoint_path)

    # Save pytorch-model
    print("Save PyTorch model to {}".format(pytorch_dump_path))
    torch.save(model.state_dict(), pytorch_dump_path)`


Haven't figured it out with the norwegian version yet though, either the model doesn't work and BotXo are just playing everyone, or I'm just not figuring it out

**Comment by kasperjunge** [comment]
> 
> 
> Where did you find the code on the left hand side?
> I am trying the same, just on the norwegian version. Does the danish vocab file start have ## before all words and ##_ before all "sub-words", kind of like sentencePiece tokenization? Appears as if the norwegian one has the same

The code on the left hand side is from the original BERT Repo:  https://github.com/google-research/bert

**Comment by kasperjunge** [comment]
> 
> 
> Found the code btw. If you just want to convert from tensorflow checkpoints into pytorch then you can just do something like this:
> 
> `from transformers import BertConfig, BertForPreTraining, load_tf_weights_in_bert`
> 
> `def convert_to_pytorch(tf_checkpoint_path, bert_config_file, pytorch_dump_path):`
> 
> ```
> # Initialise PyTorch model
> config = BertConfig.from_json_file(bert_config_file)
> print("Building PyTorch model from configuration: {}".format(str(config)))
> model = BertForPreTraining(config)
> 
> # Load weights from tf checkpoint
> load_tf_weights_in_bert(model, config, tf_checkpoint_path)
> 
> # Save pytorch-model
> print("Save PyTorch model to {}".format(pytorch_dump_path))
> torch.save(model.state_dict(), pytorch_dump_path)`
> ```
> 
> Haven't figured it out with the norwegian version yet though, either the model doesn't work and BotXo are just playing everyone, or I'm just not figuring it out

# I came up with a similar solution.  
1. I downloaded the danish pretrained BERT which contained the following files:
```
bert_config.json
bert_model.ckpt.data-00000-of-00001
bert_model.ckpt.index
bert_model.ckpt.meta
vocab.txt
```

2. Used the transformers-cli to convert model to pytorch:
```
transformers-cli convert --model_type bert --tf_checkpoint ./bert-base-danish/bert_model.ckpt --config ./bert-base-danish/bert_config.json --pytorch_dump_output ./bert-base-danish/pytorch_model.bin
```

3. Renamed bert_config.json to config.json (don't know if this is necessary, but it did not work when the config file was called bert_config.json :p)

4. Run the following code:
```
from transformers import BertConfig, BertTokenizer, BertModel

# Init BERT configuration object
configuration = BertConfig.from_json_file('./bert-base-danish/config.json')

# Init BERT tokenizer
tokenizer = BertTokenizer(vocab_file='./bert-base-danish/vocab.txt')

# Init BERT model from pretrained
model = BertModel(config=configuration).from_pretrained('bert-base-danish')
```

**Comment by christianbv** [comment]
Yes, you have to do it in that fashion (and the huggingface convention is using config.json, bert_config.json is perhaps the google convention when they initially released bert in 2018, I think). 
Good luck with your danish model! The Norwegian one released by botxo is crooked; the vocab file is entirely different and the model keeps predicting the same all the time. Wouldn't rely too much on the danish model either, to be honest

**Comment by kasperjunge** [comment]
I am doing a 2 class classification problem with the danish pre-trained model, and it does not seem to learn anything. I am currently debugging my training setup to figure out if I messed up somewhere. I will let you know if manage to make it learn somthing.
Are you sure that you are training it properly?

**Comment by kasperjunge** [comment]
Lol, I just trained the Danish pre-trained BERT model on the IMDB dataset ( https://www.kaggle.com/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews/activity ) which is in English, and I have an accuracy of 73.5% after one epoch 😂

**Comment by christianbv** [comment]
Hi again! Cool to see that it is working :) I am able to make the multilingual variant from BERT work, getting 0.81 in f1 (98% acc, 16 labels), just not the norwegian one. I haven't tested the one you're using, but i guess it's working. I believe the norwegian model don't work based on the following:

1. When creating a pipeline and masking words on the pre-trained model, the model keeps predicting the UNK token no matter what
2. Fine-tuning the model also doesn't work 

As said, my method highly depends upon the checkpoints being properly instantiated, but I don't think that is the cause
